### PR TITLE
Fix Table issue #754 and add support for replaying logged requests

### LINKED ISSUE #754: [Logic Apps] Batch API fails to update all the rows in the request
### Which service(blob, file, queue, table) does this issue concern?
table

### Which version of the Azurite was used?
`table` branch (2cf4deae207c070473a7b0be2bfe47c81a8df5b2)

### Where do you get Azurite? (npm, DockerHub, NuGet, Visual Studio Code Extension)
Local `npm` build

### What's the Node.js version?
v14.16.0

### What problem was encountered?
I did the following sequence of operations via batch APIs.
Create rows
Update rows
Delete rows

Please follow the lifecycle of row with partitionKey `09CEE` and rowKey `EA5F528CF1B84658A5CECC574848547B_FLOWIDENTIFIER-5539F65E020B44FCA32CF9CBE56E286A` in the attached log file.

It gets created at `2021-04-15T23:09:46.597Z` with tracking id `de79bba2-949c-42fe-8228-a0df1c8c8ca5`. In the response, we get eTag `W/\"datetime'2021-04-15T23%3A09%3A46.5910000Z'`.

We update the row at `2021-04-15T23:09:47.423Z` with tracking Id `90ad0628-eae6-4735-805f-9f165f0ed245`. In the response, we get a warning though. The update request contains 3 rows but only 1 is updated.

```
2021-04-15T23:09:47.422Z 90ad0628-eae6-4735-805f-9f165f0ed245 info: TableStorageContextMiddleware: RequestMethod=POST RequestURL=http://127.0.0.1/devstoreaccount1/$batch?sv=2018-03-28&sig=WTksSlJMYweYOjn0f%2BQ3AzgZlTiFEo9R9Z%2Bv8XnYXYM%3D&se=2023-04-15T23%3A09%3A41Z&srt=sco&ss=bqt&sp=racupwdl&timeout=15 RequestHeaders:{"user-agent":"ResourceStack/6.0.0.1260","x-ms-version":"2018-03-28","x-ms-client-request-id":"f2503371-15c7-4314-9803-81ea69f1ca72","accept":"application/json;odata=nometadata","content-type":"multipart/mixed; boundary=batch_3e8c6583-146e-4326-835f-5f7321fc6711","host":"127.0.0.1:10002","content-length":"6865"} ClientIP=127.0.0.1 Protocol=http HTTPVersion=1.1
```

```
2021-04-15T23:09:47.423Z 90ad0628-eae6-4735-805f-9f165f0ed245 debug: TableHandler:batch() Raw request string is "--batch_3e8c6583-146e-4326-835f-5f7321fc6711\r\nContent-Type: multipart/mixed; boundary=changeset_3e8c6583-146e-4326-835f-5f7321fc6711\r\n\r\n--changeset_3e8c6583-146e-4326-835f-5f7321fc6711\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nPOST http://127.0.0.1:10002/devstoreaccount1/flows HTTP/1.1\r\nAccept: application/json;odata=nometadata\r\nContent-Type: application/json\r\nPrefer: return-no-content\r\n\r\n{\"PartitionKey\":\"09CEE\",\"RowKey\":\"EA5F528CF1B84658A5CECC574848547B_FLOWVERSION-5539F65E020B44FCA32CF9CBE56E286A-08585830786980800821\",\"Sku\":\"{\\\"name\\\":\\\"Standard\\\",\\\"plan\\\":{\\\"name\\\":\\\"farm0\\\",\\\"id\\\":\\\"/subscriptions/ea5f528c-f1b8-4658-a5ce-cc574848547b/resourcegroups/rgname/providers/microsoft.web/serverfarms/farm0\\\",\\\"type\\\":\\\"Microsoft.Web/ServerFarms\\\"}}\",\"State\":\"Enabled\",\"CreatedTime\":\"2021-04-15T23:09:47.4209193Z\",\"CreatedTime@odata.type\":\"Edm.DateTime\",\"ChangedTime\":\"2021-04-15T23:09:47.4214726Z\",\"ChangedTime@odata.type\":\"Edm.DateTime\",\"DeletedTime\":\"1970-01-01T00:00:00Z\",\"DeletedTime@odata.type\":\"Edm.DateTime\",\"ChangedOperationId\":\"f2503371-15c7-4314-9803-81ea69f1ca72\",\"FlowId\":\"5539f65e020b44fca32cf9cbe56e286a\",\"SubscriptionId\":\"ea5f528c-f1b8-4658-a5ce-cc574848547b\",\"ResourceGroupName\":\"de415c09-29bb-483d-9544-25602c1ff355\",\"FlowName\":\"testflow1\",\"FlowSequenceId\":\"08585830786980800821\",\"ScaleUnit\":\"CU03\",\"Location\":\"devfabric\",\"RuntimeConfiguration\":\"{}\",\"DefinitionCompressed\":\"jwcotS/9AEgVAwCiRhYYkLcNkP8mJJYMoNGjbJnZPa29JAj6mJmxrTE9/2R9ohB1/NkjGtPz5ue0veuIO/Bh4F5oChYGDVOK3MToeBoFf7AtSEKrwCZGxxFRXsjU9Y9ObxOj44BjAQEAHmQzCg==\",\"DefinitionCompressed@odata.type\":\"Edm.Binary\",\"Metadata\":\"F3t9\",\"Metadata@odata.type\":\"Edm.Binary\",\"ParametersCompressed\":\"F3t9\",\"ParametersCompressed@odata.type\":\"Edm.Binary\",\"ConnectionReferences\":\"F3t9\",\"ConnectionReferences@odata.type\":\"Edm.Binary\",\"WorkflowReferences\":\"F3t9\",\"WorkflowReferences@odata.type\":\"Edm.Binary\",\"KeyVaultCertificateReferences\":\"F3t9\",\"KeyVaultCertificateReferences@odata.type\":\"Edm.Binary\",\"RuntimeContext\":\"9wwotS/9AEjlAwByiBkYkMU5BL8arW5Jf8FNkkBfmOlff9ZsdxcWrTF0mx+8MGnqWiyshMHAwMidxtAJrOk2bmWAjRV8Ta7rVnzgkVCFZ7LZJJFlmzuNoTuuSiEhqMI7jaHb/GBF15nC6V6cLr66kv9NxwIGAL9DUfY8vlXeBHXArwqTKlQZ\",\"RuntimeContext@odata.type\":\"Edm.Binary\",\"FlowUpdatedTime\":\"2021-04-15T23:09:47.4101988Z\",\"FlowUpdatedTime@odata.type\":\"Edm.DateTime\"}\r\n--changeset_3e8c6583-146e-4326-835f-5f7321fc6711\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nPUT http://127.0.0.1:10002/devstoreaccount1/flows(PartitionKey='09CEE',RowKey='EA5F528CF1B84658A5CECC574848547B_FLOWIDENTIFIER-5539F65E020B44FCA32CF9CBE56E286A') HTTP/1.1\r\nAccept: application/json;odata=nometadata\r\nContent-Type: application/json\r\nIf-Match: W/\"datetime'2021-04-15T23%3A09%3A46.5910000Z'\"\r\n\r\n{\"Sku\":\"{\\\"name\\\":\\\"Standard\\\",\\\"plan\\\":{\\\"name\\\":\\\"farm0\\\",\\\"id\\\":\\\"/subscriptions/ea5f528c-f1b8-4658-a5ce-cc574848547b/resourcegroups/rgname/providers/microsoft.web/serverfarms/farm0\\\",\\\"type\\\":\\\"Microsoft.Web/ServerFarms\\\"}}\",\"State\":\"Enabled\",\"CreatedTime\":\"2021-04-15T23:09:46.5446473Z\",\"CreatedTime@odata.type\":\"Edm.DateTime\",\"ChangedTime\":\"2021-04-15T23:09:47.4214726Z\",\"ChangedTime@odata.type\":\"Edm.DateTime\",\"DeletedTime\":\"1970-01-01T00:00:00Z\",\"DeletedTime@odata.type\":\"Edm.DateTime\",\"ChangedOperationId\":\"f2503371-15c7-4314-9803-81ea69f1ca72\",\"FlowId\":\"5539f65e020b44fca32cf9cbe56e286a\",\"SubscriptionId\":\"ea5f528c-f1b8-4658-a5ce-cc574848547b\",\"ResourceGroupName\":\"de415c09-29bb-483d-9544-25602c1ff355\",\"FlowName\":\"testflow1\",\"FlowSequenceId\":\"08585830786980800821\",\"ScaleUnit\":\"CU03\",\"Location\":\"devfabric\",\"RuntimeConfiguration\":\"{}\",\"DefinitionCompressed\":\"jwcotS/9AEgVAwCiRhYYkLcNkP8mJJYMoNGjbJnZPa29JAj6mJmxrTE9/2R9ohB1/NkjGtPz5ue0veuIO/Bh4F5oChYGDVOK3MToeBoFf7AtSEKrwCZGxxFRXsjU9Y9ObxOj44BjAQEAHmQzCg==\",\"DefinitionCompressed@odata.type\":\"Edm.Binary\",\"Metadata\":\"F3t9\",\"Metadata@odata.type\":\"Edm.Binary\",\"ParametersCompressed\":\"F3t9\",\"ParametersCompressed@odata.type\":\"Edm.Binary\",\"ConnectionReferences\":\"F3t9\",\"ConnectionReferences@odata.type\":\"Edm.Binary\",\"WorkflowReferences\":\"F3t9\",\"WorkflowReferences@odata.type\":\"Edm.Binary\",\"KeyVaultCertificateReferences\":\"F3t9\",\"KeyVaultCertificateReferences@odata.type\":\"Edm.Binary\",\"RuntimeContext\":\"9wwotS/9AEjlAwByiBkYkMU5BL8arW5Jf8FNkkBfmOlff9ZsdxcWrTF0mx+8MGnqWiyshMHAwMidxtAJrOk2bmWAjRV8Ta7rVnzgkVCFZ7LZJJFlmzuNoTuuSiEhqMI7jaHb/GBF15nC6V6cLr66kv9NxwIGAL9DUfY8vlXeBHXArwqTKlQZ\",\"RuntimeContext@odata.type\":\"Edm.Binary\",\"FlowUpdatedTime\":\"2021-04-15T23:09:47.4101988Z\",\"FlowUpdatedTime@odata.type\":\"Edm.DateTime\"}\r\n--changeset_3e8c6583-146e-4326-835f-5f7321fc6711\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nPUT http://127.0.0.1:10002/devstoreaccount1/flows(PartitionKey='09CEE',RowKey='EA5F528CF1B84658A5CECC574848547B_FLOWLOOKUP-DE415C09%3A2D29BB%3A2D483D%3A2D9544%3A2D25602C1FF355-TESTFLOW1') HTTP/1.1\r\nAccept: application/json;odata=nometadata\r\nContent-Type: application/json\r\nIf-Match: W/\"datetime'2021-04-15T23%3A09%3A46.5910000Z'\"\r\n\r\n{\"Sku\":\"{\\\"name\\\":\\\"Standard\\\",\\\"plan\\\":{\\\"name\\\":\\\"farm0\\\",\\\"id\\\":\\\"/subscriptions/ea5f528c-f1b8-4658-a5ce-cc574848547b/resourcegroups/rgname/providers/microsoft.web/serverfarms/farm0\\\",\\\"type\\\":\\\"Microsoft.Web/ServerFarms\\\"}}\",\"State\":\"Enabled\",\"CreatedTime\":\"2021-04-15T23:09:46.5446473Z\",\"CreatedTime@odata.type\":\"Edm.DateTime\",\"ChangedTime\":\"2021-04-15T23:09:47.4214726Z\",\"ChangedTime@odata.type\":\"Edm.DateTime\",\"DeletedTime\":\"1970-01-01T00:00:00Z\",\"DeletedTime@odata.type\":\"Edm.DateTime\",\"ChangedOperationId\":\"f2503371-15c7-4314-9803-81ea69f1ca72\",\"FlowId\":\"5539f65e020b44fca32cf9cbe56e286a\",\"SubscriptionId\":\"ea5f528c-f1b8-4658-a5ce-cc574848547b\",\"ResourceGroupName\":\"de415c09-29bb-483d-9544-25602c1ff355\",\"FlowName\":\"testflow1\",\"FlowSequenceId\":\"08585830786980800821\",\"ScaleUnit\":\"CU03\",\"Location\":\"devfabric\",\"RuntimeConfiguration\":\"{}\",\"DefinitionCompressed\":\"jwcotS/9AEgVAwCiRhYYkLcNkP8mJJYMoNGjbJnZPa29JAj6mJmxrTE9/2R9ohB1/NkjGtPz5ue0veuIO/Bh4F5oChYGDVOK3MToeBoFf7AtSEKrwCZGxxFRXsjU9Y9ObxOj44BjAQEAHmQzCg==\",\"DefinitionCompressed@odata.type\":\"Edm.Binary\",\"Metadata\":\"F3t9\",\"Metadata@odata.type\":\"Edm.Binary\",\"ParametersCompressed\":\"F3t9\",\"ParametersCompressed@odata.type\":\"Edm.Binary\",\"ConnectionReferences\":\"F3t9\",\"ConnectionReferences@odata.type\":\"Edm.Binary\",\"WorkflowReferences\":\"F3t9\",\"WorkflowReferences@odata.type\":\"Edm.Binary\",\"KeyVaultCertificateReferences\":\"F3t9\",\"KeyVaultCertificateReferences@odata.type\":\"Edm.Binary\",\"RuntimeContext\":\"9wwotS/9AEjlAwByiBkYkMU5BL8arW5Jf8FNkkBfmOlff9ZsdxcWrTF0mx+8MGnqWiyshMHAwMidxtAJrOk2bmWAjRV8Ta7rVnzgkVCFZ7LZJJFlmzuNoTuuSiEhqMI7jaHb/GBF15nC6V6cLr66kv9NxwIGAL9DUfY8vlXeBHXArwqTKlQZ\",\"RuntimeContext@odata.type\":\"Edm.Binary\",\"FlowUpdatedTime\":\"2021-04-15T23:09:47.4101988Z\",\"FlowUpdatedTime@odata.type\":\"Edm.DateTime\"}\r\n--changeset_3e8c6583-146e-4326-835f-5f7321fc6711--\r\n--batch_3e8c6583-146e-4326-835f-5f7321fc6711--\r\n"
```

<pre>
<b>
2021-04-15T23:09:47.424Z 	 warn: TableHandler:updateEntity() Incoming PartitionKey:09CEE RowKey:EA5F528CF1B84658A5CECC574848547B_FLOWIDENTIFIER-5539F65E020B44FCA32CF9CBE56E286A in URL parameters don't align with entity body PartitionKey:undefined RowKey:undefined.
2021-04-15T23:09:47.425Z 	 warn: TableHandler:updateEntity() Incoming PartitionKey:09CEE RowKey:EA5F528CF1B84658A5CECC574848547B_FLOWLOOKUP-DE415C09:2D29BB:2D483D:2D9544:2D25602C1FF355-TESTFLOW1 in URL parameters don't align with entity body PartitionKey:undefined RowKey:undefined.
</b>
</pre>

```
2021-04-15T23:09:47.426Z 90ad0628-eae6-4735-805f-9f165f0ed245 debug: TableHandler:batch() Raw response string is "--batchresponse_3e8c6583-146e-4326-835f-5f7321fc6711\r\nContent-Type: multipart/mixed; boundary=changesetresponse_3e8c6583-146e-4326-835f-5f7321fc6711\r\n\r\n--changesetresponse_3e8c6583-146e-4326-835f-5f7321fc6711\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nHTTP/1.1 204 No Content\r\nX-Content-Type-Options: nosniff\r\nCache-Control: no-cache\r\nLocation: http://127.0.0.1:10002/devstoreaccount1/flows(PartitionKey='09CEE',RowKey='EA5F528CF1B84658A5CECC574848547B_FLOWVERSION-5539F65E020B44FCA32CF9CBE56E286A-08585830786980800821')\r\nDataServiceId: http://127.0.0.1:10002/devstoreaccount1/flows(PartitionKey='09CEE',RowKey='EA5F528CF1B84658A5CECC574848547B_FLOWVERSION-5539F65E020B44FCA32CF9CBE56E286A-08585830786980800821')\r\nETag: W/\"datetime'2021-04-15T23%3A09%3A47.4220000Z'\"\r\n\r\n\r\n--changesetresponse_3e8c6583-146e-4326-835f-5f7321fc6711\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nHTTP/1.1 204 No Content\r\nX-Content-Type-Options: nosniff\r\nCache-Control: no-cache\r\nETag: W/\"datetime'2021-04-15T23%3A09%3A47.4220000Z'\"\r\n\r\n\r\n--changesetresponse_3e8c6583-146e-4326-835f-5f7321fc6711\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nHTTP/1.1 204 No Content\r\nX-Content-Type-Options: nosniff\r\nCache-Control: no-cache\r\nETag: W/\"datetime'2021-04-15T23%3A09%3A47.4220000Z'\"\r\n\r\n\r\n--changesetresponse_3e8c6583-146e-4326-835f-5f7321fc6711--\r\n--batchresponse_3e8c6583-146e-4326-835f-5f7321fc6711--\r\n"
```

However, on subsequent `GET` at `2021-04-15T23:09:47.451Z` with tracking id `59625f4c-8fb0-4b02-be43-dae752e16b19`, we get a new eTag for this row `"W/\\\"datetime'2021-04-15T23%3A09%3A47.4220000Z'`.

When we try to `DELETE` this row with the new eTag at `2021-04-15T23:09:47.527Z` with tracking Id `8fd74d11-29d7-4944-bf65-fdb3738e8b87`, we get `412`.

<pre>
2021-04-15T23:09:47.527Z 8fd74d11-29d7-4944-bf65-fdb3738e8b87 debug: TableHandler:batch() Raw request string is "--batch_558d985f-491c-496d-b4a2-311c3e1e075d\r\nContent-Type: multipart/mixed; boundary=changeset_558d985f-491c-496d-b4a2-311c3e1e075d\r\n\r\n--changeset_558d985f-491c-496d-b4a2-311c3e1e075d\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nDELETE http://127.0.0.1:10002/devstoreaccount1/flows(<b>PartitionKey='09CEE',RowKey='EA5F528CF1B84658A5CECC574848547B_FLOWIDENTIFIER-5539F65E020B44FCA32CF9CBE56E286A'</b>) HTTP/1.1\r\nAccept: application/json;odata=nometadata\r\nContent-Type: application/json\r\n<b>If-Match: W/\"datetime'2021-04-15T23%3A09%3A47.4220000Z'\"</b>\r\n\r\n--changeset_558d985f-491c-496d-b4a2-311c3e1e075d\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nDELETE http://127.0.0.1:10002/devstoreaccount1/flows(PartitionKey='09CEE',RowKey='EA5F528CF1B84658A5CECC574848547B_FLOWLOOKUP-DE415C09%3A2D29BB%3A2D483D%3A2D9544%3A2D25602C1FF355-TESTFLOW1') HTTP/1.1\r\nAccept: application/json;odata=nometadata\r\nContent-Type: application/json\r\nIf-Match: W/\"datetime'2021-04-15T23%3A09%3A47.4220000Z'\"\r\n\r\n--changeset_558d985f-491c-496d-b4a2-311c3e1e075d--\r\n--batch_558d985f-491c-496d-b4a2-311c3e1e075d--\r\n"
</pre>

```
2021-04-15T23:09:47.528Z 8fd74d11-29d7-4944-bf65-fdb3738e8b87 error: ErrorMiddleware: ErrorName=StorageError ErrorMessage="The update condition specified in the request was not satisfied."  ErrorHTTPStatusCode=412 ErrorHTTPStatusMessage=undefined ErrorHTTPHeaders={"x-ms-error-code":"UpdateConditionNotSatisfied","x-ms-request-id":"8fd74d11-29d7-4944-bf65-fdb3738e8b87","x-ms-version":"2019-12-12"} ErrorHTTPBody="{\"odata.error\":{\"code\":\"UpdateConditionNotSatisfied\",\"message\":{\"lang\":\"en-US\",\"value\":\"The update condition specified in the request was not satisfied.\\nRequestId:8fd74d11-29d7-4944-bf65-fdb3738e8b87\\nTime:2021-04-15T23:09:47.527Z\"}}}" ErrorStack="\"StorageError: The update condition specified in the request was not satisfied.\\n    at Function.getPreconditionFailed (C:\\\\Source\\\\Azurite\\\\dist\\\\src\\\\table\\\\errors\\\\StorageErrorFactory.js:64:16)\\n    at LokiTableMetadataStore.deleteTableEntity (C:\\\\Source\\\\Azurite\\\\dist\\\\src\\\\table\\\\persistence\\\\LokiTableMetadataStore.js:303:57)\\n    at TableHandler.deleteEntity (C:\\\\Source\\\\Azurite\\\\dist\\\\src\\\\table\\\\handlers\\\\TableHandler.js:258:34)\\n    at TableBatchOrchestrator.handleBatchDelete (C:\\\\Source\\\\Azurite\\\\dist\\\\src\\\\table\\\\batch\\\\TableBatchOrchestrator.js:253:45)\\n    at TableBatchOrchestrator.routeAndDispatchBatchRequest (C:\\\\Source\\\\Azurite\\\\dist\\\\src\\\\table\\\\batch\\\\TableBatchOrchestrator.js:138:58)\\n    at TableBatchOrchestrator.submitRequestsToHandlers (C:\\\\Source\\\\Azurite\\\\dist\\\\src\\\\table\\\\batch\\\\TableBatchOrchestrator.js:62:53)\\n    at TableBatchOrchestrator.processBatchRequestAndSerializeResponse (C:\\\\Source\\\\Azurite\\\\dist\\\\src\\\\table\\\\batch\\\\TableBatchOrchestrator.js:42:20)\\n    at TableHandler.batch (C:\\\\Source\\\\Azurite\\\\dist\\\\src\\\\table\\\\handlers\\\\TableHandler.js:400:50)\\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)\""
```

### Steps to reproduce the issue?
Please see above.

### Have you found a mitigation/solution?
No
[logicAppCRUD.log](https://github.com/Azure/Azurite/files/6321599/logicAppCRUD.log)

### PR DESCRIPTION
**Pull request by edwin-huber**
Fixes #754
Add support for replaying logged requests for easier repro of issues.
Added test case for issue #745

### CODE REVIEW

File: tests/table/apis/table.entity.tests.azure.data-tables.ts [hunk 1]
```diff
@@ -101,4 +102,93 @@ describe("table Entity APIs test", () => {
     assert.ok(result.subResponses[0].rowKey);
     await tableClient.delete();
   });
+
+  // https://github.com/Azure/Azurite/issues/754
+  it.only("Batch API should correctly process LogicApp style update request sequence", async () => {
```
Comment (XiaoningLiu) [inline tests/table/apis/table.entity.tests.azure.data-tables.ts:107]: `it.only` ==> `it`
```suggestion
  it("Batch API should correctly process LogicApp style update request sequence", async () => {
```

File: tests/table/apis/table.entity.tests.utils.for.rest.ts [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+const key1 = Buffer.from(TableEntityTestConfig.sharedKey, "base64");
+
+/**
+ * Creates the Axios request options using shared key lite
+ * Authorization header for Azurite table request
+ *
+ * @export
+ * @param {string} url
+ *  * @param {string} path
```
Comment (XiaoningLiu) [inline tests/table/apis/table.entity.tests.utils.for.rest.ts:16]:
```suggestion
 * @param {string} path
```

**Review by XiaoningLiu — APPROVED** [review]
Great work Edwin! Some minor comments added.

File: tests/table/apis/table.entity.tests.azure.data-tables.ts (hunk 1 above)
Comment (edwin-huber) [reply tests/table/apis/table.entity.tests.azure.data-tables.ts]: done

File: tests/table/apis/table.entity.tests.utils.for.rest.ts (hunk 2 above)
Comment (edwin-huber) [reply tests/table/apis/table.entity.tests.utils.for.rest.ts]: done

**Pull request merged**